Add building variants to module, derive trait for all variants and marker trait for all variants

kaleidawave opened the pull request:
Found this crate a while back and it's very cool! Had a use case for it today but needed some extra features, so here's my PR for them!

This PR adds 3 features:

#### Build variants to a module (#5)

As #5 suggests:
```rs
#[derive(EnumVariantType)]
#[evt(module = "example")]`
enum Example {
    A,
    B,
}
```

generates all the variants into a module
```rs
mod example {
     ...
}
```

Which is really useful to prevent naming conflicts and for general structure. 

#### Derive for all variants (#6)

```rs
#[derive(EnumVariantType)]
#[evt(derive(Debug, Clone))]`
enum Example {
    A,
    B,
}
```
Each generated variant has `#[derive(Debug, Clone))]`
```rs
#[derive(Debug, Clone)]
struct A;
#[derive(Debug, Clone)]
struct B;
```

Useful for generating the variants on large enums with many variants and not having to write out the derive attribute on each variant

#### Generate and derive marker trait over variants

```rs
#[derive(EnumVariantType)]
#[evt(derive_marker_trait(MyExampleMarker))]`
enum Example {
    A,
    B,
}
```

Generates a marker trait and implements it for each generated variant

```rs
trait MyExampleMarker {}
...
impl MyExampleMarker for A {}
impl MyExampleMarker for B {}
```

This is useful for constraining structures to only use the generated variants, e.g. `struct MyExampleHolder<T: MyExampleMarker>(T)`

--------

None of the features break existing uses, they only add optional things. I have added tests for each feature. Let me know if there are any changes/alterations/considerations to do...? If/when everything is good I can add the features to the documentation.

Thanks :)

azriel91 commented on src/lib.rs:
```diff
@@ -174,8 +176,70 @@ fn enum_variant_type_impl(ast: DeriveInput) -> proc_macro2::TokenStream {
+                    }
+                }
+            } else {
+                panic!("Invalid evt attr")
```
hm, it's been a while since I've done proc macros, but does rust filter out attributes that aren't listed in the `proc_macro_derive`?

if so, then we can do something like this:

```rust
panic!("`EnumVariantType` does not recognize `{}` as an attribute. Did you mean `evt`?")
```

and if not, we should probably just ignore unrecognized attributes -- I remember something about "proc macros can see all other attributes except `derive`s"

kaleidawave replied:
Interesting point, just checked and no rust does not filter attributes not specified in the attribute argument on the `proc_macro_derive`. For:

```rust
#[derive(Debug, EnumVariantType)]
#[evt(...)]
#[rustfmt::ignore]
struct A;
```
The following attributes are present

```
ast.attrs = [
    Attribute {
        path: Path {
            leading_colon: None,
            segments: [
                PathSegment {
                    ident: Ident {
                        ident: "evt",
        ...
    },
    Attribute {
        path: Path {
            leading_colon: None,
            segments: [
                PathSegment {
                    ident: Ident {
                        ident: "rustfmt",
        ...
    },
]
```

And I think this should already be handled as there isn't anything happening on not "evt" attributes. The panic is only if the evt is not a meta list e.g. `#[evt]` or `#[evt = "invalid_use"]`
```rust
for attr in ast.attrs.iter() {
    if attr.path.is_ident("evt") {
        if let Ok(Meta::List(list)) = attr.parse_meta() {
            ...
        } else {
            panic!("Invalid evt attr")
        }
    }
}
```

I couldn't think of a better error message 😄 Maybe something which references correct usage or links to it. Also maybe should be returning a `compile_error!` macro rather than panicking. Will have a look at tidying my error logic up 👍

azriel91 replied:
> And I think this should already be handled as there isn't anything happening on not "evt" attributes. The panic is only if the evt is not a meta list e.g. #[evt] or #[evt = "invalid_use"]

oh yes, my eyes didn't parse that right (though this was at the 2-indent level) :bowing_man:

azriel91 commented on src/lib.rs:
```diff
@@ -174,8 +176,70 @@ fn enum_variant_type_impl(ast: DeriveInput) -> proc_macro2::TokenStream {
+                                marker_trait_ident = Some(ident);
+                            }
+                        }
+                        _ => panic!("Invalid evt attr"),
```
similar to the other comment, can add suggestions of what's valid

azriel91 commented on src/lib.rs:
```diff
@@ -174,8 +176,70 @@ fn enum_variant_type_impl(ast: DeriveInput) -> proc_macro2::TokenStream {
+                                    #[derive( #(#items),* )]
+                                });
+                            } else if list.path.is_ident("derive_marker_trait") {
+                                if list.nested.len() != 1 {
```
eh, what if, someone wants to derive multiple marker traits?

with this guard they can do this :smile:

```rust
#[evt(derive_marker_trait(TraitA))]
#[evt(derive_marker_trait(TraitB))]
```

might as well allow them the shorthand

kaleidawave replied:
Haha I was debating this. Was wondering about what use case would require two marker traits given that the traits and implementoreers would be exactly the same? Decided it was more of anti-pattern and not to permit it. Will change to allow for a shorthand version if there is a good reason...?

azriel91 replied:
hm, use case I can think of is marker traits from two separate crates

other way round, I can't think of many cases where preventing it protects the user very much -- specifying the same marker trait in an attribute should be rare, and even if they do, the rust compilation error would probably be clear enough (multiple conflicting implementations)

kaleidawave replied:
Ah there is a misunderstanding. `#[evt(derive_marker_trait(MyTrait))` actually generates a new marker trait `MyTrait` definition (e.g. it generates `trait MyTrait {}`, it is not a reference to a existing trait just the name of a new one to generate.
https://github.com/kaleidawave/enum_variant_type/blob/1908e53bbaf7930bc388328235e285470eb0fb36/src/lib.rs#L645

However your comment here is actually very good, I can think of some use cases where I would want to implement existing marker traits to all variants. And if I wrote the `MyTrait` trait myself and it didn't generate it, it would have the same effect. I think I will modify the name to `implement_marker_traits`, take the arguments as being references and remove the one argument guard (as I agree makes no sense if implementing over existing 😄 )

azriel91 replied:
ah! my mistake -- didn't realize the marker trait was also generated

> I think I will modify the name to implement_marker_traits, take the arguments as being references and remove the one argument guard (as I agree makes no sense if implementing over existing smile )

cool :v:; and ya the 1 argument guard makes sense in the "let's generate a marker trait" case :stuck_out_tongue:

azriel91 commented on src/lib.rs:
```diff
@@ -433,4 +515,179 @@ mod tests {
+            #[evt(derive(Debug))]
+            pub enum MyEnum {
+                A,
+                B
```
probably worth adding an inner derive to `B`, and asserting we have both `#[derive(Debug)]` and `#[derive(Clone)]` or something. we don't have to be clever and merge them into `#[derive(Clone, Debug)]` -- i don't think people will mind having

```rust
#[derive(Debug)]
#[derive(Clone)]
pub struct B;
```

generated, and there's no technical disadvantage

kaleidawave replied:
Yep that's a great thing to add into the tests! And that's actually the currently implementation to *stack* the derives. I learnt when adding this feature that rust allows for this and handles having attributes with the same path on the same item which makes this alot easier :)

azriel91 reviewed (commented):
heya :wave:! Thank you very much for the contribution; am happy to put them in.

The code and tests are very good :bow:, I've added some improvements for the error messages, and one thing that I wasn't sure it's safe to panic on

kaleidawave commented:
Awesome following the conversations have changed the `implement_marker_traits` thing, improved some error messages, changed some of the tests and have added documentation for the new features. I think I am confident with these changes and they are ready for merging! Feel free if there are any other things todo or you can reword any of the documentation I added. 

Wa just re-read your messages and I think you may changed some errors messages but not pushed the changes, feel free to rewrite them over mine.

azriel91 reviewed (commented):
cool, shall merge soon -- I haven't actually written the error message for that outermost case, but shall do that now

azriel91 commented:
ah, it doesn't pick up .github/actions/ci.yml because it lives on the fork. Shall merge and see

azriel91 commented:
cool, got the other random tidy up work done (probably should've done it in its own branch, but ya)

thank you very much for implementing this to a high standard :bowing_man:

oh, it's released in [`0.3.0`](https://crates.io/crates/enum_variant_type/0.3.0) :rocket:.

kaleidawave commented:
Awesome, thanks for the quick responses and adding my changes to 0.3.0! 🚀 🚀
